### PR TITLE
add dind.enable easy-button and 'podContainers' param to allow specification of agent side-cars

### LINKED ISSUE #54: Buildkite should provide Docker-in-Docker functionalities
This ticket is related to the following issues happening when using Buildkite on Kubernetes:
- https://github.com/buildkite-plugins/docker-buildkite-plugin/issues/61
- https://github.com/buildkite-plugins/docker-buildkite-plugin/issues/140
- https://github.com/buildkite/agent/issues/729

The following configuration works successfully on Google Kubernetes Engine:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: buildkite-agent 
  namespace: buildkite
spec:
  containers:
    - name: agent
      image: buildkite/agent:3.17.0
      imagePullPolicy: Always
      env: 
      - name: BUILDKITE_AGENT_TOKEN
        valueFrom:
          secretKeyRef:
            key: agent-token
            name: buildkite-agent-agent
      - name: BUILDKITE_AGENT_META_DATA
        value: role=agent
      - name: BUILDKITE_BUILD_PATH
        value: /var/buildkite/builds
      - name: DOCKER_HOST
        value: tcp://localhost:2375
      volumeMounts:
        - name: shared-volume
          mountPath: /var/buildkite
    - name: dind
      image: docker:19.03-dind
      securityContext:
        privileged: true
      env:
        - name: DOCKER_TLS_CERTDIR
          value: ""
      volumeMounts:
        - name: docker-graph-storage
          mountPath: /var/lib/docker
        - name: shared-volume
          mountPath: /var/buildkite
  volumes:
    - name: docker-graph-storage
      emptyDir: {}
    - name: shared-volume
      emptyDir: {}
```

With the following pipeline:
```
steps:
  - commands:
      - "docker build ."
    plugins:
      - docker#v3.3.0:
          image: "docker:latest"
          always-pull: true
          volumes:
            - "/var/run/docker.sock:/var/run/docker.sock"
```

I spent around 16 hours trying to figure out how to work around GKE security limitations and how to make Buildkite work with Docker-out-of-Docker, but I ended up reaching the same conclusion I did for Drone CI: Docker-in-Docker is more suitable for the Kubernetes architecture **especially** for Google Kubernetes Engine architecture.

The following article also explains really well why Docker-in-Docker (dind) is considered more functional in a Kubernetes environment:
https://applatix.com/case-docker-docker-kubernetes-part-2/

The main highlight from the article is the following:

> The Pod will create a container that will run outside of the Pod. By running the container using DooD, you lose out on the following for the spawned container:
> - Pod Networking - Cannot access the container using Pod IP.
> - Pod Lifecycle - On Pod termination, this container will keep running especially if the container was started with -d flag.
> - Pod Cleanup - Graph storage will not be cleanup after pod terminates.
> - Scheduling and Resource Utilization - Cpu and Memory requested by Pod, will only be for the Pod and not the container spawned from the Pod. Also, limits on CPU and memory set for the Pod will not be inherited by the spawned container.

Aside from that, as far as I know the following tools are all using dind instead of dood when deployed to a k8s environment:
- Gitlab CI
- Drone CI
- Jenkins CI
- GoCD CI

It also looks like AWS is no longer working with DooD without additional configuration (i.e. Buildkite won't work out-of-the-box on ECS without the `--enable-docker-bridge` flag):
https://support.cloudbees.com/hc/en-us/articles/360028151031-Docker-outside-of-Docker-no-longer-works-in-EKS

**This is also the only configuration I am aware of that fully runs on GKE without breaking down when the docker container is used.**

### PR DESCRIPTION
**Pull request by O1ahmad**
**What this PR does / why we need it**:

This PR enables the injection of **"podContainers"** or sidecars into *Buildkite Agent (bk-agent)* pod environments for enabling chart operators with the ability to provide auxiliary and enhanced environment capabilities for a variety of use-cases. Example uses include:
- CI/CD job environment isolation utilizing the Docker-in-Docker or **"DinD"** sidecar technique
- Prometheus Node Exporter as a sidecar for monitoring individual buildkite agent performance statistics/characteristics
- Jaegar sidecar daemon for collecting application traces during build, test and other types of workflows

**Which issue this PR fixes** *(optional, in `fixes` format, will close that issue when PR gets merged)*:

fixes #54 

**Special notes for your reviewer**:

This change is meant to be as minimal and general as possible in order to support as wide-range of sidecar use possible without adding significant code overhead for general functionality or error-checking in specification. 

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [X] Chart Version bumped
- [X] Variables are documented in the README.md

### CODE REVIEW

**Comment by O1ahmad** [comment]
**# Example: DinD -podContainer-**

```yaml
$ cat Values.yaml
---
# buildkite agent settings
image:
  repository: buildkite/agent
agent:
  token: "<token>"
  meta: "<tags>"
enableHostDocker: false
extraEnv:
  - name: DOCKER_HOST
    value: "tcp://localhost:2375"
volumeMounts:
  - name: shared-volume
    mountPath: /var/buildkite

# additional Pod container or sidecars specifications
podContainers:
  - name: dind
    image: "docker:19.03-dind"
    securityContext:
      privileged: true
    env:
      - name: DOCKER_TLS_CERTDIR
        value: ""
    volumeMounts:
      - name: "docker-graph-storage"
        mountPath: "/var/lib/docker"
      - name: shared-volume
        mountPath: "/var/buildkite"

# available volumes across Pod
volumes:
  - name: docker-graph-storage
    emptyDir: {}
  - name: shared-volume
    emptyDir: {}
```

**Comment by christian-roggia** [comment]
I like the minimal-impact solution but I think it would be a lot more user-friendly to implement a preconfigured DinD configuration (something that you might want to turn on in the values.yaml with a `dind.enabled=true`).

Users normally get scared when they need to paste so many lines (they probably don't fully understand) in the values.yaml.

Either way, I think this is a step in the right direction.

**Comment by O1ahmad** [comment]
Hey, thanks!..and agreed for the most part. I like the more user-friendly option for a preconfigured **DinD** configuration like with `dind.enabled=true` also.

How about allowing for the general `podContainers` spec (like with *volumes*, *extraEnvs*), providing those interested with an assortment of sidecars to be expressed, in addition to the `dind` switch (w/ perhaps a few knobs for controlling the `dind image` and `daemon REST listening port`)?

**Comment by christian-roggia** [comment]
I don't see anything wrong with providing more flexibility even though I believe it's ultimately the owners' decision whether they want to include it or not. I personally fully support your approach to the issue.

**Comment by rimusz** [comment]
@0x0I thanks for the PR, the optional `DinD` support looks good 👍

**Comment by rimusz** [comment]
new chart version got released

File: stable/agent/templates/deployment.yaml [hunk 1]
```diff
@@ -81,6 +88,25 @@ spec:
             - name: docker-config
               mountPath: /root/.docker
             {{- end }}
+            {{- if .Values.dind.enabled }}
+            - name: shared-volume
+              mountPath: "/var/buildkite"
```
Comment (tjwallace) [inline stable/agent/templates/deployment.yaml:93]: Should this be `/buildkite`? (the [default location for docker](https://buildkite.com/docs/agent/v3/docker#default-file-locations))

**Pull request merged**